File: app/users/rpc/lanes.go
package rpc

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
	"go.uber.org/zap"
)

// Lane telemetry RPC. A "lane" is one JetStream durable (or ephemeral)
// consumer created by the fleet's subscribers (see pkg/bus). The retired Go
// admin read this directly over its own NATS connection and rendered it; the
// SvelteKit console cannot speak the JetStream management API (its client is a
// core request/reply helper), so this service answers on its behalf.
//
// Every number is read live from the broker — nothing is simulated. The data
// path is StreamsInfo -> per-stream ConsumersInfo -> laneWire, sampled by a
// background ticker so each `.get` reply carries a real two-sample throughput
// without re-walking the JetStream API per request.
//
// Subjects live under the existing admin-user prefix (bagel.rpc.admin.user.*):
// the console's `admin` NATS user already publishes that wildcard and the
// `users` service user already subscribes it and holds $JS.> permission, so the
// lane view needs no broker-auth change to come online.

// laneWire is the exact wire contract the console's LaneView type expects. Field
// names and JSON shape must not drift from console/admin/src/lib/server/lanes.ts.
type laneWire struct {
	Stream      string `json:"stream"`
	Consumer    string `json:"consumer"`
	Display     string `json:"display"`
	Subject     string `json:"subject"`
	Category    string `json:"category"` // "system" | "projection" | "ephemeral"
	Ephemeral   bool   `json:"ephemeral"`
	Orphan      bool   `json:"orphan"`
	Pending     uint64 `json:"pending"`
	InFlight    string `json:"inFlight"`
	Rate        string `json:"rate"`
	Redelivered int    `json:"redelivered"`
}

type lanesReply struct {
	Lanes []laneWire `json:"lanes"`
	Error string     `json:"error,omitempty"`
}

type laneMutationReply struct {
	OK     bool   `json:"ok"`
	Notice string `json:"notice,omitempty"`
	Error  string `json:"error,omitempty"`
}

type laneMutationRequest struct {
	Stream   string `json:"stream"`
	Consumer string `json:"consumer"`
	Alias    string `json:"alias"`
}

// SubscribeLanes wires the lane telemetry + mutation RPC on the admin prefix and
// starts the background sampler. The sampler runs until ctx is cancelled.
func SubscribeLanes(ctx context.Context, nc *nats.Conn, prefix, queueGroup string, log *zap.Logger) error {
	js, err := nc.JetStream()
	if err != nil {
		return fmt.Errorf("lanes: jetstream context: %w", err)
	}

	s := &laneSampler{js: js, store: newLaneStore(js), prev: map[string]laneSample{}, log: log}
	s.sample() // prime one observation so a rate appears on the next tick
	go s.run(ctx)

	handlers := map[string]nats.MsgHandler{
		"lanes.get": func(msg *nats.Msg) {
			lanes, errMsg := s.snapshot()
			respondJSON(msg, lanesReply{Lanes: lanes, Error: errMsg})
		},
		"lanes.alias":   s.handleAlias,
		"lanes.durable": s.handleDurable,
		"lanes.delete":  s.handleDelete,
	}

	for verb, handle := range handlers {
		subject := prefix + "." + verb
		if _, err := nc.QueueSubscribe(subject, queueGroup, handle); err != nil {
			return fmt.Errorf("lanes: subscribe %s: %w", subject, err)
		}
	}
	return nil
}

func respondJSON(msg *nats.Msg, v any) {
	body, _ := json.Marshal(v)
	_ = msg.Respond(body)
}

// laneSample is one prior observation of a consumer's delivery counter.
type laneSample struct {
	delivered uint64
	at        time.Time
}

// laneSampler holds the previous delivery sample per (stream, consumer) so it
// can compute a real per-second throughput across two ticks, plus the latest
// rendered snapshot. It is shared across requests behind a mutex.
type laneSampler struct {
	js    nats.JetStreamContext
	store *laneStore
	log   *zap.Logger

	mu      sync.Mutex
	prev    map[string]laneSample
	value   []laneWire
	lastErr string
}

func laneKey(stream, consumer string) string { return stream + "\x00" + consumer }

// run samples on a 2s cadence (the same window the retired admin polled at) so
// any `.get` reply has a fresh two-sample rate, until ctx is cancelled.
func (s *laneSampler) run(ctx context.Context) {
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			s.sample()
		}
	}
}

// snapshot returns the last rendered lanes and error string.
func (s *laneSampler) snapshot() ([]laneWire, string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.value, s.lastErr
}

// sample walks the JetStream API once, recomputes throughput against the prior
// observation, merges the operator's display aliases, and stores the result.
func (s *laneSampler) sample() {
	lanes, errMsg := s.collect(time.Now())
	s.mu.Lock()
	s.value = lanes
	s.lastErr = errMsg
	s.mu.Unlock()
}

// collect enumerates every stream and its consumers, diffs each consumer's
// Delivered.Consumer counter against the prior sample for a real throughput, and
// renders the wire rows. A non-empty error string means the JetStream API is
// unreachable or this account lacks $JS.API access.
//
// It deliberately does NOT probe AccountInfo first: the hub runs JetStream with
// a domain (domain: hub in nats-server.conf), so a bare $JS.API.INFO request has
// no responder and AccountInfo times out, even though the stream/consumer list
// calls below ARE served on the bare API. Reachability is inferred from whether
// any stream comes back instead.
func (s *laneSampler) collect(now time.Time) ([]laneWire, string) {
	aliases := s.store.aliases()

	type raw struct {
		stream      string
		consumer    string
		filter      string
		ephemeral   bool
		orphan      bool
		category    string
		group       string
		pending     uint64
		ackPending  int
		maxAckPend  int
		redelivered int
		delivered   uint64
		rate        float64
		hasRate     bool
	}

	var rows []raw
	seen := map[string]struct{}{}
	streamsSeen := 0

	for stream := range s.js.StreamsInfo() {
		if stream == nil {
			continue
		}
		streamsSeen++
		streamName := stream.Config.Name
		for ci := range s.js.ConsumersInfo(streamName) {
			if ci == nil {
				continue
			}
			key := laneKey(streamName, ci.Name)
			seen[key] = struct{}{}

			filter := ci.Config.FilterSubject
			ephemeral := ci.Config.Durable == ""
			r := raw{
				stream:      streamName,
				consumer:    ci.Name,
				filter:      filter,
				ephemeral:   ephemeral,
				orphan:      !ci.PushBound,
				category:    laneCategory(streamName, ephemeral),
				group:       laneGroup(ci.Name, filter, ephemeral),
				pending:     ci.NumPending,
				ackPending:  ci.NumAckPending,
				maxAckPend:  ci.Config.MaxAckPending,
				redelivered: ci.NumRedelivered,
				delivered:   ci.Delivered.Consumer,
			}

			// Throughput from two real samples. No prior sample => no rate yet.
			if prev, ok := s.prev[key]; ok {
				if secs := now.Sub(prev.at).Seconds(); secs > 0 {
					delta := float64(ci.Delivered.Consumer) - float64(prev.delivered)
					if delta < 0 {
						// Consumer was reset/recreated and the monotonic counter
						// went backwards; a negative rate is meaningless.
						delta = 0
					}
					r.rate = delta / secs
					r.hasRate = true
				}
			}
			s.prev[key] = laneSample{delivered: ci.Delivered.Consumer, at: now}

			rows = append(rows, r)
		}
	}

	// StreamsInfo's channel closes empty both when the JetStream API is
	// unreachable (no responder / no permission) and when the broker genuinely
	// has no streams. The fleet always provisions BAGEL_DATA, TWITCH_INGRESS and
	// TWITCH_OUTGRESS (see pkg/bus DataStreams), so zero streams here means this
	// account could not read the JetStream API, not an empty broker.
	if streamsSeen == 0 {
		return nil, "JetStream API unreachable: no streams returned (broker unreachable or account lacks $JS.API access)"
	}

	// Drop stale samples for consumers that no longer exist so the map does not
	// grow without bound across recreations.
	for key := range s.prev {
		if _, ok := seen[key]; !ok {
			delete(s.prev, key)
		}
	}

	sortRows := func(i, j int) bool {
		a, b := rows[i], rows[j]
		if ra, rb := categoryRank(a.category), categoryRank(b.category); ra != rb {
			return ra < rb
		}
		if a.stream != b.stream {
			return a.stream < b.stream
		}
		if a.filter != b.filter {
			return a.filter < b.filter
		}
		return a.consumer < b.consumer
	}
	sort.SliceStable(rows, sortRows)

	lanes := make([]laneWire, 0, len(rows))
	for _, r := range rows {
		alias := aliases[laneAliasKey(r.stream, r.consumer)]
		lanes = append(lanes, laneWire{
			Stream:      r.stream,
			Consumer:    r.consumer,
			Display:     displayName(alias, r.group, r.consumer, r.ephemeral),
			Subject:     r.filter,
			Category:    r.category,
			Ephemeral:   r.ephemeral,
			Orphan:      r.orphan,
			Pending:     r.pending,
			InFlight:    inFlightText(r.ackPending, r.maxAckPend),
			Rate:        rateText(r.rate, r.hasRate),
			Redelivered: r.redelivered,
		})
	}
	return lanes, ""
}

func (s *laneSampler) handleAlias(msg *nats.Msg) {
	var req laneMutationRequest
	if err := json.Unmarshal(msg.Data, &req); err != nil {
		respondJSON(msg, laneMutationReply{Error: "bad request"})
		return
	}
	if req.Stream == "" || req.Consumer == "" {
		respondJSON(msg, laneMutationReply{Error: "rename failed: missing lane"})
		return
	}
	alias := strings.TrimSpace(req.Alias)
	if len(alias) > 48 {
		alias = alias[:48]
	}
	if err := s.store.setAlias(req.Stream, req.Consumer, alias); err != nil {
		s.log.Warn("lane alias failed", zap.String("consumer", req.Consumer), zap.Error(err))
		respondJSON(msg, laneMutationReply{Error: "rename failed: " + err.Error()})
		return
	}
	s.sample() // reflect the new label on the next read immediately
	if alias == "" {
		respondJSON(msg, laneMutationReply{OK: true, Notice: "alias cleared"})
		return
	}
	respondJSON(msg, laneMutationReply{OK: true, Notice: "renamed to " + alias})
}

// handleDurable converts an ephemeral lane into a permanent durable consumer
// with the same subject filter. Nothing drains it, so it retains new messages
// until the stream's own retention reclaims them. Refused on already-durable
// lanes.
func (s *laneSampler) handleDurable(msg *nats.Msg) {
	var req laneMutationRequest
	if err := json.Unmarshal(msg.Data, &req); err != nil {
		respondJSON(msg, laneMutationReply{Error: "bad request"})
		return
	}
	info, err := s.js.ConsumerInfo(req.Stream, req.Consumer)
	if err != nil {
		respondJSON(msg, laneMutationReply{Error: "make-permanent failed: " + err.Error()})
		return
	}
	if info.Config.Durable != "" {
		respondJSON(msg, laneMutationReply{Error: "lane is already durable"})
		return
	}
	name := "adminperm_" + subjectToken(info.Config.FilterSubject)
	if _, err := s.js.AddConsumer(req.Stream, &nats.ConsumerConfig{
		Durable:       name,
		Description:   "operator-pinned permanent lane (admin)",
		FilterSubject: info.Config.FilterSubject,
		AckPolicy:     nats.AckExplicitPolicy,
		DeliverPolicy: nats.DeliverNewPolicy,
	}); err != nil {
		s.log.Warn("lane make-durable failed", zap.String("consumer", req.Consumer), zap.Error(err))
		respondJSON(msg, laneMutationReply{Error: "make-permanent failed: " + err.Error()})
		return
	}
	s.log.Info("lane made permanent", zap.String("stream", req.Stream),
		zap.String("source", req.Consumer), zap.String("durable", name))
	s.sample()
	respondJSON(msg, laneMutationReply{OK: true,
		Notice: "created permanent lane " + name + " (nothing drains it; it retains until stream retention)"})
}

// handleDelete removes a consumer with no bound subscriber ("no system
// attached"). It refuses any lane whose deliver subject still has a live
// subscriber, so a briefly-restarting service is never deleted out from under
// itself.
func (s *laneSampler) handleDelete(msg *nats.Msg) {
	var req laneMutationRequest
	if err := json.Unmarshal(msg.Data, &req); err != nil {
		respondJSON(msg, laneMutationReply{Error: "bad request"})
		return
	}
	info, err := s.js.ConsumerInfo(req.Stream, req.Consumer)
	if err != nil {
		respondJSON(msg, laneMutationReply{Error: "delete failed: " + err.Error()})
		return
	}
	if info.PushBound {
		respondJSON(msg, laneMutationReply{Error: "refused: lane is bound to a running consumer, not an orphan"})
		return
	}
	if err := s.js.DeleteConsumer(req.Stream, req.Consumer); err != nil {
		s.log.Warn("lane delete failed", zap.String("consumer", req.Consumer), zap.Error(err))
		respondJSON(msg, laneMutationReply{Error: "delete failed: " + err.Error()})
		return
	}
	_ = s.store.setAlias(req.Stream, req.Consumer, "") // drop any stale alias
	s.log.Info("lane deleted", zap.String("stream", req.Stream), zap.String("consumer", req.Consumer))
	s.sample()
	respondJSON(msg, laneMutationReply{OK: true, Notice: "deleted orphan lane " + req.Consumer})
}

// displayName is the human label: the operator alias wins, then "ephemeral" for
// the per-pod broadcast subscribers whose real names are random gibberish, then
// the service/group for a durable consumer, then the raw name.
func displayName(alias, group, consumer string, ephemeral bool) string {
	if alias != "" {
		return alias
	}
	if ephemeral {
		return "ephemeral"
	}
	if group != "" {
		return group
	}
	return consumer
}

// inFlightText renders the in-flight (delivered, unacked) count, with the cap
// when the consumer bounds it.
func inFlightText(ackPending, maxAckPend int) string {
	if maxAckPend > 0 {
		return fmt.Sprintf("%d / %d", ackPending, maxAckPend)
	}
	return fmt.Sprint(ackPending)
}

// rateText renders throughput. The first time a lane is seen there is no prior
// sample, so it shows a dash rather than a fabricated number.
func rateText(rate float64, hasRate bool) string {
	if !hasRate {
		return "—"
	}
	switch {
	case rate == 0:
		return "0 msg/s"
	case rate < 10:
		return fmt.Sprintf("%.1f msg/s", rate)
	default:
		return fmt.Sprintf("%.0f msg/s", math.Round(rate))
	}
}

// laneGroup derives the service/queue group from a durable consumer name. The
// fleet names durables <group>_<subjectToken(filter)> (see pkg/bus durableName),
// so stripping the tokenised filter suffix recovers the group. Ephemeral
// consumers have random names and no meaningful group.
func laneGroup(name, filter string, ephemeral bool) string {
	if ephemeral {
		return ""
	}
	if filter != "" {
		if g := strings.TrimSuffix(name, "_"+subjectToken(filter)); g != name {
			return g
		}
	}
	return name
}

// subjectToken mirrors pkg/bus.subjectToken: dots and wildcards become
// underscores so the value is usable inside a JetStream consumer name.
func subjectToken(subject string) string {
	return strings.NewReplacer(".", "_", "*", "_", ">", "_").Replace(subject)
}

// laneCategory buckets a lane for sectioning in the UI. The outgress stream
// carries the chat egress "system" lanes; other durable consumers fold
// data-plane events (projections); ephemeral consumers are the per-pod
// cache-invalidation broadcast subscribers.
func laneCategory(stream string, ephemeral bool) string {
	switch {
	case ephemeral:
		return "ephemeral"
	case stream == "TWITCH_OUTGRESS":
		return "system"
	default:
		return "projection"
	}
}

// categoryRank orders the sections: system lanes first (the ones operators
// watch), then projections, then the ephemeral noise.
func categoryRank(category string) int {
	switch category {
	case "system":
		return 0
	case "projection":
		return 1
	default:
		return 2
	}
}

File: console/admin/src/lib/server/lanes.ts
import { jsm, js } from '@bagel/shared/server/nats';
import type { KV } from 'nats';
import { dev } from '$app/environment';

// Keep this boot-safe. Importing access.ts here pulls in services.ts, creating
// a cycle through hooks.server.ts while adapter-node awaits initialization.
// Branching on `dev` in this module also lets Rollup erase the fixture import.
const DEMO = dev && process.env.DEMO === '1';

export interface LaneView {
  stream: string;
  consumer: string;
  display: string;
  subject: string;
  category: string;
  ephemeral: boolean;
  orphan: boolean;
  pending: number;
  inFlight: string;
  rate: string;
  redelivered: number;
}

export interface LanesResult {
  lanes: LaneView[];
  degraded: boolean;
  notice: string;
}

export interface LaneMutationResult {
  ok: boolean;
  notice?: string;
  error?: string;
}

const LANE_BUCKET = 'admin_lanes';
const LANE_STREAM = `KV_${LANE_BUCKET}`;
const NATS_REPLICAS = 3;
let kvStore: KV | null = null;

async function getKV(): Promise<KV> {
  if (kvStore) return kvStore;
  const client = await js();
  const manager = await jsm();
  try {
    const info = await manager.streams.info(LANE_STREAM);
    if (info.config.num_replicas !== NATS_REPLICAS) {
      await manager.streams.update(LANE_STREAM, {
        ...info.config,
        num_replicas: NATS_REPLICAS
      });
    }
    kvStore = await client.views.kv(LANE_BUCKET, { history: 1, replicas: NATS_REPLICAS });
  } catch (err: any) {
    if (err.code === '404' || err.message?.includes('not found')) {
      kvStore = await client.views.kv(LANE_BUCKET, {
        history: 1,
        replicas: NATS_REPLICAS,
        description: 'admin lane display aliases'
      });
    } else {
      throw err;
    }
  }
  return kvStore;
}

// Best-effort boot reconciliation; request-time callers retry through getKV if
// the hub did not yet have quorum while this process was starting.
export async function ensureLaneStoreHA(): Promise<void> {
  await getKV();
}

function laneKey(stream: string, consumer: string) {
  return `${stream}\x00${consumer}`;
}

function laneAliasKey(stream: string, consumer: string) {
  return `${stream}.${consumer}`;
}

interface LaneSample {
  delivered: number;
  at: number;
}

const prevSamples = new Map<string, LaneSample>();
let currentLanes: LaneView[] = [];
let lastError = '';
let samplerTimer: ReturnType<typeof setInterval> | null = null;
let sampling: Promise<void> | null = null;
let lastLoadAt = 0;

const SAMPLE_INTERVAL_MS = 5_000;
const SAMPLE_IDLE_MS = 60_000;
const ALIAS_CACHE_TTL_MS = 30_000;

let aliasCache = new Map<string, string>();
let aliasCacheExpires = 0;

function subjectToken(subject: string) {
  return subject.replace(/[.*>]/g, '_');
}

function laneGroup(name: string, filter: string, ephemeral: boolean) {
  if (ephemeral) return '';
  if (filter) {
    const token = subjectToken(filter);
    if (name.endsWith(`_${token}`)) {
      return name.slice(0, -token.length - 1);
    }
  }
  return name;
}

function laneCategory(stream: string, ephemeral: boolean) {
  if (ephemeral) return 'ephemeral';
  // Both outgress streams are the egress/control lanes: TWITCH_OUTGRESS (chat)
  // and TWITCH_OUTGRESS_SYSTEM (EventSub enroll + stream_status).
  if (stream.startsWith('TWITCH_OUTGRESS')) return 'system';
  return 'projection';
}

function categoryRank(category: string) {
  switch (category) {
    case 'system': return 0;
    case 'projection': return 1;
    default: return 2;
  }
}

function displayName(alias: string | undefined, group: string, consumer: string, ephemeral: boolean) {
  if (alias) return alias;
  if (ephemeral) return 'ephemeral';
  if (group) return group;
  return consumer;
}

function inFlightText(ackPending: number, maxAckPend: number) {
  if (maxAckPend > 0) return `${ackPending} / ${maxAckPend}`;
  return `${ackPending}`;
}

function rateText(rate: number, hasRate: boolean) {
  if (!hasRate) return '—';
  if (rate === 0) return '0 msg/s';
  if (rate < 10) return `${rate.toFixed(1)} msg/s`;
  return `${Math.round(rate)} msg/s`;
}

function markAliasesDirty() {
  aliasCacheExpires = 0;
}

async function loadAliases(): Promise<Map<string, string>> {
  if (aliasCacheExpires > Date.now()) return aliasCache;

  const kv = await getKV();
  const aliases = new Map<string, string>();
  try {
    // Collect keys first, then fetch values in parallel (chunked so a large
    // alias set can't fan out unbounded). The old per-key serial `await kv.get`
    // paid one KV round-trip per alias — 100 aliases = 100 sequential hops.
    const keys: string[] = [];
    const keysIter = await kv.keys();
    for await (const k of keysIter) keys.push(k);
    const CHUNK = 50;
    for (let i = 0; i < keys.length; i += CHUNK) {
      const chunk = keys.slice(i, i + CHUNK);
      const entries = await Promise.all(chunk.map((k) => kv.get(k)));
      entries.forEach((e, j) => {
        if (e) aliases.set(chunk[j], e.string());
      });
    }
  } catch (err: any) {
    if (err.code !== '404') console.warn('lane alias fetch error:', err);
  }

  aliasCache = aliases;
  aliasCacheExpires = Date.now() + ALIAS_CACHE_TTL_MS;
  return aliasCache;
}

async function collectLanes() {
  if (sampling) return sampling;
  sampling = collectLanesOnce().finally(() => {
    sampling = null;
  });
  return sampling;
}

interface LaneRow {
  stream: string;
  consumer: string;
  filter: string;
  ephemeral: boolean;
  orphan: boolean;
  category: string;
  group: string;
  pending: number;
  ackPending: number;
  maxAckPend: number;
  redelivered: number;
  rate: number;
  hasRate: boolean;
}

// One consumer-list round trip per stream, all in flight at once. The old
// serial walk paid streams x consumers sequential JS API hops, which is what
// made the lanes page feel stuck. allSettled keeps partial data when a single
// stream's listing fails; the failure is surfaced, not hidden.
async function listStreamConsumers(manager: Awaited<ReturnType<typeof jsm>>) {
  const streams = await manager.streams.list().next();
  const listed = await Promise.allSettled(
    streams.map(async (stream) => ({
      streamName: stream.config.name,
      consumers: await manager.consumers.list(stream.config.name).next()
    }))
  );
  return {
    streamsSeen: streams.length,
    failures: listed.filter((r) => r.status === 'rejected').length,
    fulfilled: listed
      .filter((r) => r.status === 'fulfilled')
      .map((r) => (r as PromiseFulfilledResult<{ streamName: string; consumers: any[] }>).value)
  };
}

// sampleRate derives msg/s from the delivered-sequence delta since the last
// pass and stores the new baseline.
function sampleRate(key: string, deliveredSeq: number, now: number): { rate: number; hasRate: boolean } {
  const prev = prevSamples.get(key);
  prevSamples.set(key, { delivered: deliveredSeq, at: now });
  if (!prev) return { rate: 0, hasRate: false };
  const secs = (now - prev.at) / 1000;
  if (secs <= 0) return { rate: 0, hasRate: false };
  return { rate: Math.max(deliveredSeq - prev.delivered, 0) / secs, hasRate: true };
}

function laneRowOf(streamName: string, ci: any, now: number): LaneRow {
  const filter = ci.config.filter_subject || '';
  const ephemeral = !ci.config.durable_name;
  const { rate, hasRate } = sampleRate(laneKey(streamName, ci.name), ci.delivered.consumer_seq, now);
  return {
    stream: streamName,
    consumer: ci.name,
    filter,
    ephemeral,
    orphan: !ci.push_bound,
    category: laneCategory(streamName, ephemeral),
    group: laneGroup(ci.name, filter, ephemeral),
    pending: ci.num_pending,
    ackPending: ci.num_ack_pending,
    maxAckPend: ci.config.max_ack_pending || 0,
    redelivered: ci.num_redelivered,
    rate,
    hasRate
  };
}

function compareLanes(a: LaneRow, b: LaneRow): number {
  return (
    categoryRank(a.category) - categoryRank(b.category) ||
    a.stream.localeCompare(b.stream) ||
    a.filter.localeCompare(b.filter) ||
    a.consumer.localeCompare(b.consumer)
  );
}

function laneViewOf(r: LaneRow, aliases: Map<string, string>): LaneView {
  return {
    stream: r.stream,
    consumer: r.consumer,
    display: displayName(aliases.get(laneAliasKey(r.stream, r.consumer)), r.group, r.consumer, r.ephemeral),
    subject: r.filter,
    category: r.category,
    ephemeral: r.ephemeral,
    orphan: r.orphan,
    pending: r.pending,
    inFlight: inFlightText(r.ackPending, r.maxAckPend),
    rate: rateText(r.rate, r.hasRate),
    redelivered: r.redelivered
  };
}

// Only prune rate baselines on a complete listing: after a partial one, a
// missing key means "stream unreadable this pass", not "consumer gone".
function pruneStaleBaselines(seen: Set<string>) {
  for (const key of prevSamples.keys()) {
    if (!seen.has(key)) prevSamples.delete(key);
  }
}

async function collectLanesOnce() {
  try {
    const manager = await jsm();
    const aliases = await loadAliases();
    const { streamsSeen, failures, fulfilled } = await listStreamConsumers(manager);

    if (streamsSeen === 0) {
      lastError = "JetStream API unreachable: no streams returned (broker unreachable or account lacks $JS.API access)";
      return;
    }

    const now = Date.now();
    const rows = fulfilled.flatMap(({ streamName, consumers }) =>
      consumers.map((ci) => laneRowOf(streamName, ci, now))
    );
    if (failures === 0) {
      pruneStaleBaselines(new Set(rows.map((r) => laneKey(r.stream, r.consumer))));
    }

    rows.sort(compareLanes);
    currentLanes = rows.map((r) => laneViewOf(r, aliases));
    lastError = failures > 0 ? `partial listing: ${failures} of ${streamsSeen} streams unreadable` : '';
  } catch (err: any) {
    lastError = err.message || String(err);
  }
}

function ensureSampler() {
  lastLoadAt = Date.now();
  if (samplerTimer) return;
  samplerTimer = setInterval(() => {
    if (Date.now() - lastLoadAt > SAMPLE_IDLE_MS) {
      if (samplerTimer) clearInterval(samplerTimer);
      samplerTimer = null;
      return;
    }
    collectLanes();
  }, SAMPLE_INTERVAL_MS);
}

export async function loadLanes(): Promise<LanesResult> {
  if (DEMO) {
    const { sampleLanes } = await import('./demo-data');
    return { lanes: sampleLanes, degraded: false, notice: '' };
  }
  ensureSampler();
  // Cold cache: wait for the in-flight collection instead of returning an
  // empty list that pretends to be the fleet. Warm cache serves instantly and
  // refreshes in the background.
  const pending = collectLanes();
  if (currentLanes.length === 0) await pending;
  if (lastError) {
    return {
      lanes: currentLanes,
      degraded: true,
      notice: 'Lane telemetry error: ' + lastError
    };
  }
  return { lanes: currentLanes, degraded: false, notice: '' };
}

export async function laneAlias(stream: string, consumer: string, alias: string): Promise<LaneMutationResult> {
  try {
    const kv = await getKV();
    const key = laneAliasKey(stream, consumer);
    if (!alias) {
      await kv.delete(key);
      markAliasesDirty();
      collectLanes(); // sample immediately
      return { ok: true, notice: 'alias cleared' };
    }
    await kv.put(key, new TextEncoder().encode(alias.slice(0, 48)));
    markAliasesDirty();
    collectLanes(); // sample immediately
    return { ok: true, notice: 'renamed to ' + alias };
  } catch (err: any) {
    return { ok: false, error: 'rename failed: ' + err.message };
  }
}

export async function laneDurable(stream: string, consumer: string): Promise<LaneMutationResult> {
  try {
    const manager = await jsm();
    const info = await manager.consumers.info(stream, consumer);
    if (info.config.durable_name) {
      return { ok: false, error: 'lane is already durable' };
    }
    const name = "adminperm_" + subjectToken(info.config.filter_subject || '');
    await manager.consumers.add(stream, {
      ...info.config,
      durable_name: name,
      description: "operator-pinned permanent lane (admin)"
    });
    collectLanes();
    return { ok: true, notice: `created permanent lane ${name}` };
  } catch (err: any) {
    return { ok: false, error: 'make-permanent failed: ' + err.message };
  }
}

export async function laneDelete(stream: string, consumer: string): Promise<LaneMutationResult> {
  try {
    const manager = await jsm();
    const info = await manager.consumers.info(stream, consumer);
    if (info.push_bound) {
      return { ok: false, error: 'refused: lane is bound to a running consumer, not an orphan' };
    }
    await manager.consumers.delete(stream, consumer);
    const kv = await getKV();
    await kv.delete(laneAliasKey(stream, consumer)).catch(() => {});
    markAliasesDirty();
    collectLanes();
    return { ok: true, notice: `deleted orphan lane ${consumer}` };
  } catch (err: any) {
    return { ok: false, error: 'delete failed: ' + err.message };
  }
}
